### PR TITLE
Modeling pipeline to finetune Llama-3.2-3B-Instruct using the esconv …

### LINKED ISSUE #6: Deep learning approach
This is an open source LLM that's fine tuned for our chosen domain (cognitive behavior therapy/emotional support). We will evaluate this approach against the naive baseline (a generic language model with no fine tuning, ideally the same foundation model) and the traditional ML approach.

### PR DESCRIPTION
**Pull request by haranku16**
…dataset, quantize, and upload to Huggingface Hub

# Add Deep Learning Approach with Fine-tuned Emotional Support Model

This PR implements a deep learning pipeline that fine-tunes the Llama 3.2 3B model on emotional support conversations. The model was trained on Google Colab over 5 epochs using torchtune with LoRA (Low-Rank Adaptation) to efficiently adapt the model for empathetic responses based on Cognitive Behavioral Therapy principles.

### Data Processing
- Extract emotional support conversation dataset from Hugging Face
- Transform raw data into OpenAI chat format with system instructions
- Validate data format for proper sequence of messages

### Model Training
- Configure LoRA parameters for efficient fine-tuning of Llama 3.2 3B
- Apply fine-tuning to attention modules with rank 64
- Optimize using AdamW with cosine learning rate schedule

### Model Deployment
- Export and quantize the fine-tuned model using llama.cpp
- Reduce model size with q5_k_m quantization for efficient deployment
- Upload quantized model to Hugging Face Hub for sharing

### Usage Example
- Include example script for invoking the quantized model
- Configure proper context and parameters for optimal performance

**Output: That sounds like a tough situation. What is going on?**

## Requirements

- **CMake** is required to build llama.cpp for model quantization
- Python dependencies as listed in the updated requirements.txt
- Hugging Face account and access token

## Getting Started
1. Copy `.env_example` to `.env` and add your Hugging Face credentials
2. Install dependencies from the updated requirements.txt
3. Run `pipeline.sh` to execute the full workflow, or run individual scripts for specific steps

This approach delivers an emotionally supportive AI assistant that can apply principles of Cognitive Behavioral Therapy in conversations while maintaining efficiency through quantization.

## Hugging Face Hub

https://huggingface.co/haran-nallasivan/meowth-nlp-demo-0.1_llama-3.2-3b-instruct_q5_k_m_gguf

----

Closes #6

### CODE REVIEW

File: scripts/deep/validate.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
+import json
+
+def validate_openai_conversations(file_path):
+    """
+    Validates ShareGPT conversations for proper message order and role alternation.
```
Comment (haranku16) [inline scripts/deep/validate.py:5]: Approach was modified from ShareGPT format to OpenAI format. Comment needs to be adjusted accordingly.

File: scripts/deep/transform.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+def transform_to_openai(json_file_path, output_file_path):
+    """
+    Transforms a JSON data file into a ShareGPT-formatted dataset.
+
+    Args:
+        json_file_path (str): Path to the input JSON file.
+        output_file_path (str): Path to the output ShareGPT-formatted JSON file.
+    """
+
+    instruction = """You are an emotional support counselor specializing in Cognitive Behavioral Therapy (CBT). Your role is to provide empathetic and supportive guidance to users, helping them understand and manage their thoughts, feelings, and behaviors. You will use CBT techniques to help users identify negative thought patterns, challenge them, and develop healthier coping mechanisms.
```
Comment (haranku16) [inline scripts/deep/transform.py:12]: I need to add a comment that this instruction was produced through Perplexity.

**Comment by AkhilByteWrangler** [comment]
Hey @haranku16 , 
This is an awesome PR, super detailed and well-structured! You've done a great job walking through the entire pipeline from data processing to model deployment. Below are some thoughts and suggestions to make it even better.

---

### What’s Really Good?  

- **Clear Documentation** – The way you've broken down the workflow makes it easy to follow. Love the inclusion of an example response!  
- **Efficient Fine-Tuning** – Using LoRA to fine-tune Llama 3.2 3B is a smart move to keep training efficient.  
- **Quantization for Deployment** – q5_k_m quantization helps keep the model lightweight while maintaining performance.  
- **Easy Setup & Requirements** – The steps for installation and execution are well-documented. Setting things up looks straightforward!  

---

### Areas for Improvement  

#### 1. Fine-Tuning Details  
- Would love a bit more insight into **why rank 64** was chosen for LoRA.   

#### 2. Deployment Considerations  
- Does quantization introduce **trade-offs in inference quality**? If so, how noticeable is it?  

#### 3. Ethical Considerations  
- Were any **bias mitigation steps** taken? It’s good to highlight these since this is a sensitive use case.

Overall, this is a **fantastic** PR - really well thought out and practical. Addressing some of the above points would make it even stronger and give more transparency into the model’s performance and deployment trade-offs.  

Looking forward to seeing this in action! Let me know what you think - happy to discuss further! 

Best,
Akhil C (@AkhilByteWrangler)

**Review by AkhilByteWrangler — APPROVED** [review]

**Comment by johnjoel2001** [comment]
Hey @haranku16,
This is an impressive implementation! Fine-tuning Llama 3.2 3B with LoRA for empathetic responses is a great approach to balancing efficiency and performance. The use of Cognitive Behavioral Therapy principles adds significant value to the model’s emotional support capabilities. Also, leveraging `llama.cpp` for quantization ensures optimized deployment.

**A couple of quick questions**: 

1.  Have you tested response quality post-quantization? 
2.  Also, would it be beneficial to include a small benchmark comparing the fine-tuned model’s responses with the base model? 

Overall, this is a well-structured pipeline - great work!

**Comment by haranku16** [comment]
> Hey @haranku16, This is an impressive implementation! Fine-tuning Llama 3.2 3B with LoRA for empathetic responses is a great approach to balancing efficiency and performance. The use of Cognitive Behavioral Therapy principles adds significant value to the model’s emotional support capabilities. Also, leveraging `llama.cpp` for quantization ensures optimized deployment.
> 
> **A couple of quick questions**:
> 
> 1. Have you tested response quality post-quantization?
> 2. Also, would it be beneficial to include a small benchmark comparing the fine-tuned model’s responses with the base model?
> 
> Overall, this is a well-structured pipeline - great work!

Thanks for the callout @johnjoel2001. Response quality testing is already being covered as part of Ashley's workstream, but I've reached out to her to also add the base quantized model (without fine tuning) as one of the models for evaluation. @Ashley-Xu for visibility

File: scripts/deep/validate.py (hunk 1 above)
Comment (haranku16) [reply scripts/deep/validate.py:5]: To be addressed in #10

File: scripts/deep/transform.py (hunk 2 above)
Comment (haranku16) [reply scripts/deep/transform.py:12]: To be addressed in #10

**Comment by haranku16** [comment]
> Hey @haranku16 , This is an awesome PR, super detailed and well-structured! You've done a great job walking through the entire pipeline from data processing to model deployment. Below are some thoughts and suggestions to make it even better.
> 
> ### What’s Really Good?
> * **Clear Documentation** – The way you've broken down the workflow makes it easy to follow. Love the inclusion of an example response!
> * **Efficient Fine-Tuning** – Using LoRA to fine-tune Llama 3.2 3B is a smart move to keep training efficient.
> * **Quantization for Deployment** – q5_k_m quantization helps keep the model lightweight while maintaining performance.
> * **Easy Setup & Requirements** – The steps for installation and execution are well-documented. Setting things up looks straightforward!
> 
> ### Areas for Improvement
> #### 1. Fine-Tuning Details
> * Would love a bit more insight into **why rank 64** was chosen for LoRA.
> 
> #### 2. Deployment Considerations
> * Does quantization introduce **trade-offs in inference quality**? If so, how noticeable is it?
> 
> #### 3. Ethical Considerations
> * Were any **bias mitigation steps** taken? It’s good to highlight these since this is a sensitive use case.
> 
> Overall, this is a **fantastic** PR - really well thought out and practical. Addressing some of the above points would make it even stronger and give more transparency into the model’s performance and deployment trade-offs.
> 
> Looking forward to seeing this in action! Let me know what you think - happy to discuss further!
> 
> Best, Akhil C (@AkhilByteWrangler)

Thanks Akhil for the thoughtful feedback. I have opened a new issue #11 to address this, as well as other required elements to include in the README.
